File: couchbase-persistence/couchbase-persistence-scala-sbt/hello-impl/src/main/scala/com/lightbend/lagom/samples/hello/impl/HelloServiceImpl.scala
package com.lightbend.lagom.samples.hello.impl

import akka.Done
import akka.NotUsed
import akka.util.Timeout
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.cluster.sharding.typed.scaladsl.EntityRef
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.samples.hello.api.HelloService
import com.lightbend.lagom.samples.hello.api.UserGreeting
import com.lightbend.lagom.scaladsl.api.transport.BadRequest

import scala.concurrent.duration._
import scala.collection.immutable.Seq
import scala.concurrent.ExecutionContext

class HelloServiceImpl(
  greetingsRepository: GreetingsRepository,
  clusterSharding: ClusterSharding
)(implicit ec: ExecutionContext)
    extends HelloService {

  /**
   * Looks up the entity for the given ID.
   */
  private def entityRef(id: String): EntityRef[HelloCommand] =
    clusterSharding.entityRefFor(HelloState.typeKey, id)

  implicit val timeout = Timeout(5.seconds)

  override def hello(id: String): ServiceCall[NotUsed, String] = ServiceCall {
    _ =>
      // Look up the sharded entity (aka the aggregate instance) for the given ID.
      val ref = entityRef(id)

      // Ask the aggregate instance the Hello command.
      ref
        .ask[Greeting](replyTo => Hello(id, replyTo))
        .map(greeting => greeting.message)
  }

  override def useGreeting(id: String) = ServiceCall { request =>
    // Look up the sharded entity (aka the aggregate instance) for the given ID.
    val ref = entityRef(id)

    // Tell the aggregate to use the greeting message specified.
    ref
      .ask[Confirmation](
        replyTo => UseGreetingMessage(id, request.message, replyTo)
      )
      .map {
        case Accepted => Done
        case _        => throw BadRequest("Can't upgrade the greeting message.")
      }
  }
  override def userGreetings(): ServiceCall[NotUsed, Seq[UserGreeting]] =
    ServiceCall { _ =>
      greetingsRepository.listUserGreetings()
    }

}

File: couchbase-persistence/couchbase-persistence-java-mvn/hello-impl/src/main/java/com/lightbend/lagom/samples/hello/impl/HelloServiceImpl.java
package com.lightbend.lagom.samples.hello.impl;

import akka.Done;
import akka.NotUsed;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.Entity;
import akka.cluster.sharding.typed.javadsl.EntityRef;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.api.transport.BadRequest;
import com.lightbend.lagom.samples.hello.api.GreetingMessage;
import com.lightbend.lagom.samples.hello.api.HelloService;
import com.lightbend.lagom.samples.hello.api.UserGreeting;
import com.lightbend.lagom.samples.hello.impl.HelloCommand.Hello;
import com.lightbend.lagom.samples.hello.impl.HelloCommand.UseGreetingMessage;

import javax.inject.Inject;
import java.time.Duration;
import java.util.List;

public class HelloServiceImpl implements HelloService {

    private final Duration askTimeout = Duration.ofSeconds(5);
    private ClusterSharding clusterSharding;
    private GreetingsRepository greetingsRepository;

    @Inject
    public HelloServiceImpl(GreetingsRepository greetingsRepository,
                            ClusterSharding clusterSharding) {
        this.clusterSharding = clusterSharding;
        this.greetingsRepository = greetingsRepository;

        // register the Aggregate as a sharded entity
        this.clusterSharding.init(
            Entity.of(
                HelloAggregate.ENTITY_TYPE_KEY,
                HelloAggregate::create
            )
        );
    }

    @Override
    public ServiceCall<NotUsed, String> hello(String id) {
        return request -> {
            // Look up the aggregete instance for the given ID.
            EntityRef<HelloCommand> ref = clusterSharding.entityRefFor(HelloAggregate.ENTITY_TYPE_KEY, id);
            // Ask the entity the Hello command.
            return ref.
                <HelloCommand.Greeting>ask(replyTo -> new Hello(id, replyTo), askTimeout)
                .thenApply(greeting -> greeting.message);
        };
    }

    @Override
    public ServiceCall<GreetingMessage, Done> useGreeting(String id) {
        return request -> {
            // Look up the aggregete instance for the given ID.
            EntityRef<HelloCommand> ref = clusterSharding.entityRefFor(HelloAggregate.ENTITY_TYPE_KEY, id);
            // Tell the entity to use the greeting message specified.
            return ref.
                <HelloCommand.Confirmation>ask(replyTo -> new UseGreetingMessage(request.getMessage(), replyTo), askTimeout)
                .thenApply(confirmation -> {
                        if (confirmation instanceof HelloCommand.Accepted) {
                            return Done.getInstance();
                        } else {
                            throw new BadRequest(((HelloCommand.Rejected) confirmation).reason);
                        }
                    }
                );
        };
    }

    @Override
    public ServiceCall<NotUsed, List<UserGreeting>> userGreetings() {
        return request ->
            greetingsRepository.listUserGreetings();
    }

}
